src/erl_optics_foil_server.erl

```diff
@@ -0,0 +1,114 @@
+-module(erl_optics_foil_server).
+-include("erl_optics.hrl").
+
+-behaviour(gen_server).
+
+-export([
+    init/1,
+    handle_cast/2,
+    handle_call/3,
+    handle_info/2,
+    terminate/2,
+    code_change/3
+]).
+
+-export([
+    start_link/0,
+    add_lens/2,
+    stop/0
+]).
+
+-define(SERVER, ?MODULE).
+
+-record(state, {
+    foil_namespace :: ?NS,
+    new_lenses  :: map(),
+    timer_ref :: timer:tref() | undefined
+}).
+
+%%%=========
+%%% API
+%%%=========
+
+%Modes: carbon | blank
+
+start_link() ->
+    gen_server:start_link({local, ?SERVER}, ?MODULE, [], []).
+
+stop() ->
+    gen_server:cast(?SERVER, stop).
+
+%%%==========
+%%% Callbacks
+%%%==========
+
+%% TODO: Change the timer to start after a new lens has been
+%%       declared and stop after a while to avoid a permanent
+%%       server doing useless things.
+
+init(_) ->
+    Interval = ?ENV(?ENV_FOIL_RELOAD_INTERVAL, ?DEFAULT_FOIL_RELOAD_INTERVAL),
+    {ok, Tref} = timer:send_interval(Interval, reload_foil),
+    {ok, #state{new_lenses = #{}, timer_ref = Tref}, 0}.
+
+handle_cast({add_lens, Lens}, State) ->
+    {Key, Ptr} = Lens,
+    case maps:find(Key, State#state.new_lenses) of
+        {ok, _Ptr} ->
+            erl_optics_nif:lens_close(Ptr),
+            {noreply, State};
+        _ ->
+            {noreply, #state{new_lenses =  maps:put(Key, Ptr, State#state.new_lenses)}}
+    end;
+handle_cast(stop, State) ->
+    {stop, normal, State};
+handle_cast(_, State) ->
+    {noreply, State}.
+
+handle_call(_, _From, State) ->
+    {reply, {error, undefined_call}, State}.
+
+handle_info(reload_foil, State) ->
+    Lenses = check_foil(maps:to_list(State#state.new_lenses)),
+    case Lenses of
+        [] ->
+            {noreply, #state{new_lenses = #{}}};
+        _ ->
+            lists:map(fun insert_lens/1, Lenses),
+            foil:load(?NS),
+            {noreply, #state{new_lenses = #{}}}
+    end;
+handle_info(_, State) ->
+    {noreply, State}.
+
+
+%%% API
+
+add_lens(Key, Ptr) ->
+    gen_server:cast(?SERVER, {add_lens, {Key, Ptr}}).
+
+terminate(_Reason, State) ->
+    timer:cancel(State#state.timer_ref),
+    ok.
+
+code_change(_OldVsn, State, _Extra) ->
+    {ok, State}.
+
+%%% private
+
+insert_lens({Key, Ptr}) ->
+    foil:insert(?NS, Key, Ptr).
+
+check_foil([], Acc) ->
+    Acc;
+check_foil([{Key, Ptr} | Rest], Acc) ->
+    case foil:lookup(?NS, Key) of
+        {ok, _Ptr} ->
+            erl_optics_nif:lens_close(Ptr),
```

Comment:
What purpose does this serve? Are later lenses with the same key supposed to override older ones? If so, shouldn't the recursive call cons the key on to `Acc`?

Reply:
the alloc_get functions from optics return a new struct that contains a pointer to the lens. It doesn't matter which one is kept, as long as there is only one, and the other are freed. 

That being said, I think this part is going to be removed since there is now a check prior to sending it to the foil server.